src/ripple/app/misc/AMMHelpers.h

```diff
@@ -163,43 +376,114 @@ std::optional<TAmounts<TIn, TOut>>
+                auto const nTakerPaysConstraint =
+                    pool.out * quality.rate() - pool.in / f;
+                if (nTakerPaysPropose > nTakerPaysConstraint)
+                    return nTakerPaysConstraint;
+                return nTakerPaysPropose;
+            }();
+            if (nTakerPays <= 0)
+            {
+                JLOG(j.trace())
+                    << "changeSpotPriceQuality negative: " << to_string(pool.in)
+                    << " " << to_string(pool.out) << " " << quality << " "
+                    << tfee;
+                return std::nullopt;
+            }
+            auto const takerPays = toAmount<TIn>(
+                getIssue(pool.in), nTakerPays, Number::rounding_mode::upward);
+            // should not fail
+            if (auto const amounts =
+                    TAmounts<TIn, TOut>{
+                        takerPays, swapAssetIn(pool, takerPays, tfee)};
+                Quality{amounts} < quality &&
+                !withinRelativeDistance(
+                    Quality{amounts}, quality, Number(1, -7)))
+            {
+                JLOG(j.error())
+                    << "changeSpotPriceQuality failed: " << to_string(pool.in)
+                    << " " << to_string(pool.out) << " "
+                    << to_string(amounts.in) << " " << to_string(amounts.out)
+                    << " " << quality << " " << tfee;
+                Throw<std::runtime_error>("changeSpotPriceQuality failed");
+            }
+            else
+            {
+                JLOG(j.trace())
+                    << "changeSpotPriceQuality succeeded: "
+                    << to_string(pool.in) << " " << to_string(pool.out) << " "
+                    << to_string(amounts.in) << " " << to_string(amounts.out)
+                    << " " << quality << " " << tfee;
+                return amounts;
+            }
+        }
+        JLOG(j.trace()) << "changeSpotPriceQuality negative: "
+                        << to_string(pool.in) << " " << to_string(pool.out)
+                        << " " << quality << " " << tfee;
+        return std::nullopt;
+    }
+
+    // Generate the offer starting with XRP side. Return seated offer amounts
+    // if the offer can be generated, otherwise nullopt.
+    auto const amounts = [&]() {
+        if (isXRP(getIssue(pool.out)))
+            return getAMMOfferStartWithTakerGets(pool, quality, tfee);
+        return getAMMOfferStartWithTakerPays(pool, quality, tfee);
+    }();
+    if (!amounts)
+    {
+        JLOG(j.trace()) << "changeSpotPrice negative: " << to_string(pool.in)
+                        << " " << to_string(pool.out) << " "
+                        << Number{1} / quality.rate() << " " << tfee
+                        << std::endl;
+        return std::nullopt;
     }
-    return std::nullopt;
+
+    // Might fail due to finite precision. Should the relative difference be
+    // allowed?
+    if (Quality{*amounts} < quality)
```

Comment:
Since we're defining rounding to favor the AMM, won't this trigger almost all the time? I.e. I'd _expect_ the computed quality to be less than the requested quality (by tiny amounts, of course). I might be wrong on that, but if I am can you leave a comment in the code explaining why we don't expect to hit this very often?

Reply:
There are two scenario. One is - AMM offer is generated so that the new Spot Price Quality is equal to the requested quality. In this case the offer's quality is better than updated SPQ. The second is - if there is a trading fee then matching the new SPQ to the target quality will push AMM offer quality bellow the target quality. In this case the AMM offer is generated to match the target quality and SPQ quality is going to be better than the target quality. In both cases we start the offer generation with the XRP side (if applicable) and we round it downward. This minimizes the offer size and maximizes the quality. If the offer is IOU/IOU, we still round downward the side of the offer that we start generating first. This again minimizes the offer and maximizes the quality. So the scenario when we succeed in the offer generation but it's quality is less than the target quality should be pretty rare. It's still possible though and perhaps we should add some error allowance.